src/monodroid/jni/monodroid-glue.c

```diff
@@ -3098,48 +3267,67 @@ convert_dl_flags (int flags)
 static void*
 monodroid_dlopen (const char *name, int flags, char **err, void *user_data)
 {
+	int dl_flags = convert_dl_flags (flags);
+	void *h = NULL;
+	char *full_name = NULL;
+	char *basename = NULL;
+
 	/* name is NULL when we're P/Invoking __Internal, so remap to libmonodroid */
-	char *full_name = path_combine (app_libdir, name ? name : "libmonodroid.so");
-	if (!name && !file_exists (full_name)) {
-		log_info (LOG_ASSEMBLY, "Trying to load library '%s'", full_name);
-		free (full_name);
+	if (name == NULL) {
+		name = "libmonodroid.so";
+		h = load_dso_from_override_dirs (name, dl_flags);
```

Comment:
I feel like there are a lot of places in this PR where we call `load_dso_from_override_dirs()` then `load_dso_from_app_lib_dirs()` if `load_dso_from_override_dirs()` fails.

Perhaps we should have a helper which does that?  `load_dso()`, perhaps?  (Which admittedly *already* exists, but I'm seriously getting confused trying to keep all these new functions straight.)